File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#include <iostream>
#include <vector>
#include <algorithm>

using namespace std;

int K, N;
long long nLen;
vector<long long> kLen;

long long binarySearch(long long left, long long right){
    if(left > right)
        return nLen;

    long long mid = (left + right) / 2;
    int cnt = 0;

    for(int len : kLen){
        cnt += len/mid;
    }
         
    // N개 이상으로 만들어지면 범위를 더 크게 잡기
    if(cnt >= N) {
        nLen = mid;
        return binarySearch(mid + 1, right);
    }   
    // N개 미만으로 만들어지면 범위를 더 작게 잡기
    else {
    	return binarySearch(left, mid - 1);
    }
}

int main(){
    long long len;

    cin >> K >> N;
    for(int i=0; i<K; i++){
        cin >> len;
        kLen.push_back(len);
    }
    sort(kLen.begin(), kLen.end(), greater<>()); // 내림차순 정렬
    
    cout << binarySearch(1, kLen[0]);

    return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.BufferedReader;
import java.io.InputStreamReader;

public class Main {
    private static int K, N;
    private static long nLen;
    private static long[] kLen;

    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        String[] lens = br.readLine().split(" ");
        K = Integer.parseInt(lens[0]);
        N = Integer.parseInt(lens[1]);
        kLen = new long[K];

        long max = 0;
        for(int i=0; i<K; i++){
            kLen[i] = Long.parseLong(br.readLine());
            if(max < kLen[i]) max = kLen[i];
        }

        System.out.println(binarySearch(1, max));
    }
    
    private static long binarySearch(long left, long right){
        if(left <= right) {
            long mid = (left + right) / 2;
            int cnt = 0;

            for(long len : kLen){
                cnt += len / mid;
            }

            if(cnt >= N){
                nLen = mid;
                binarySearch(mid+1, right);
            }
            else
                binarySearch(left, mid-1);
        }
        return nLen;
    }
}